### PR TITLE
[WIP] Composable SFT

### PR DESCRIPTION
**Pull request by haileyschoelkopf**
https://arxiv.org/pdf/2110.07560.pdf <-- Paper
https://github.com/cambridgeltl/composable-sft <-- code

Their code is very well packaged, so it wasn't too hard to add 

The file `sft_args` is essentially their `SftArguments` class, but I've made some of the defaults closer to the lang adaptation in their paper/what we should use.

TODOs:
- make sure that logging loss works correctly. At the moment, the code from the paper seems to not log loss correctly, so I may need to edit their code to fix it here.
- Make it easier for the user to just swap `--train_adapter` and `--lang_adapt_strategy` out for `--train_sft`. This means manually overriding some parts of sft_args in the script with model_args.
- Determine hyperparameters we should use for comparable testing. This will mean, likely, 50k train steps + 50k rewinded steps with one iteration in their method. Or maybe 5 iterations + 10k train steps twice? Idk yet
- Add a computation for converting between their "K parameters to tune" vs. number of total parameters in adapters. Can hard-code a formula for this, I don't want to mess with adding then removing an adapter

If we want to train *both adapters and Composable SFT* at once, this will require some extra code. Probably not TOO bad, but would need extra testing to account for freezing all correct parameters

### CODE REVIEW

**Comment by haileyschoelkopf** [comment]
Issue #23 is addressed by this

**Comment by haileyschoelkopf** [comment]
@yongzx This is ready to merge, the only thing needing changing is the calc for # of parameters to finetune.

**Comment by yongzx** [comment]
> Yong, did you run calcs to find # of tunable adapter params, or just get that number by adding an adapter to the model?

> the only thing needing changing is the calc for # of parameters to fine-tune.

I can help do this, no worries! It's just a running sum of trainable parameters. I will review the code over the weekend.

**Comment by yongzx** [comment]
I just did a quick read of the commit. It seems like for SFT, we don't need to modify anything in the adapter-transformers?

**Comment by haileyschoelkopf** [comment]
Yep! Just 
```
git clone https://github.com/cambridgeltl/composable-sft.git
cd composable-sft
pip install -e .
```
to install their code. Thanks!

**Comment by haileyschoelkopf** [comment]
Also

> I can help do this, no worries! It's just a running sum of trainable parameters. I will review the code over the weekend.

what I meant by this was to set the number of parameters this method changes (it's fully configurable) such that that total was equivalent to using `pfeiffer+inv` adapters, not just counting # trainable params with this method.

**Comment by yongzx** [comment]
I will prioritize evaluation test suites over this for now, but I hope to finish reviewing this before our meeting this Friday.

**Comment by haileyschoelkopf** [comment]
No problem! 

For reference, 
```
git clone https://github.com/haileyschoelkopf/composable-sft.git
cd composable-sft
pip install -e .
```
Now to install the dependency, do this instead. I'm hoping to add a `Random` and `FISH` masking strategy in this.

File: scripts/lang_adapt/madx_run_clm.py [hunk 1]
```diff
@@ -105,7 +115,7 @@ class ModelArguments:
     )
     lang_adapt_strategies: str = field(
         default="",
-        metadata={"help": "choose one of the three strategies - 'emb', 'emb-and-adpt', 'emb-then-adpt'"},
+        metadata={"help": "choose one of the three strategies - 'emb', 'emb-and-adpt', 'emb-then-adpt', 'emb-and-sft'"},
```
Comment (yongzx) [inline scripts/lang_adapt/madx_run_clm.py:118]: right now I will remove `emb-and-*` as it should be the default. So `emb-and-sft` will become `sft`.

File: scripts/lang_adapt/madx_run_clm.py [hunk 2]
```diff
@@ -711,11 +834,50 @@ def main():
...
+        if adapter_args.adapter_reduction_factor:
+            logger.info(f"Overriding K value for SFT with adapter_reduction_factor: {adapter_args.adapter_reduction_factor}")
+            # calc appropriate K value
+            num_layers = len(model.transformer.h)
+            sft_k = num_layers * model.transformer.word_embeddings.weight.shape[1] ** 2 // adapter_args.adapter_reduction_factor * 2 #* 2 for the up and down proj
+
+            sft_k += model.transformer.word_embeddings.weight.shape[1]  ** 2 // 2 # inv adapters. TODO: if we use other adapter configs, this breaks (code works, but K no longer matches adapter budget)
+
+            adapter_args.ft_params_num = int(sft_k)
+            logger.info(f"K value for SFT is {adapter_args.ft_params_num}")
```
Comment (yongzx) [inline scripts/lang_adapt/madx_run_clm.py:849]: Can I move this line 838 - 849 into the condition after line 702?

File: scripts/lang_adapt/madx_run_clm.py [hunk 3]
```diff
@@ -711,11 +834,50 @@ def main():
...
+
+    if adapter_args.train_adapter:
```
Comment (yongzx) [inline scripts/lang_adapt/madx_run_clm.py:851]: Can I remove this since we are not training adapters and SFT at the same time?

File: scripts/lang_adapt/madx_run_clm.py (hunk 2 above)
Comment (haileyschoelkopf) [reply scripts/lang_adapt/madx_run_clm.py:849]: Yes! Also, does the K value computed here match what the # of adapter parameters should be?

File: scripts/lang_adapt/madx_run_clm.py (hunk 3 above)
Comment (haileyschoelkopf) [reply scripts/lang_adapt/madx_run_clm.py:851]: Lines 851-865 were just for testing if this computation for K is the same as adapter number of parameters, so they could be deleted.

File: scripts/lang_adapt/madx_run_clm.py [hunk 4]
```diff
@@ -711,11 +834,50 @@ def main():
...
+
+        num_layers = len(model.transformer.h)
+        sft_k = num_layers * model.transformer.word_embeddings.weight.shape[1] ** 2 // adapter_args.adapter_reduction_factor * 2 #* 2 for the up and down proj
+
+        sft_k += model.transformer.word_embeddings.weight.shape[1] ** 2 // 2 # inv adapters. TODO: if we use other adapter configs, this breaks (code works, but K no longer matches adapter budget)
+
+        adapter_args.ft_params_num = int(sft_k)
+        logger.info(f"K value for SFT is {adapter_args.ft_params_num}")
+    
+    # only needed for composable sft
```
[the remark this answers is not in the document]
Comment (haileyschoelkopf) [reply scripts/lang_adapt/madx_run_clm.py:867]: Yes, as long as this is done after setting all required parameter `requires_grad` to False

**Comment by yongzx** [comment]
558f674 now supports composable SFT. Hailey, do you want to test it out?

**Comment by haileyschoelkopf** [comment]
Yes, let me try running this with those parameters!
